### PR TITLE
Add Builder Codes

### PR DESCRIPTION
**Pull request by Frixoe**
I've added the initial builder codes approval code having referenced it from the python sdk and I think I'm missing something in the implementation related to my understanding of the builders code API.

When I tried running it initially I got a 422 and now I get a 403 but that's probably a personal environment issue. But would be great to get some feedback on the current implementation. Please ignore the info! lines.

### CODE REVIEW

File: src/bin/approve_builder_fee.rs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+    let exchange_client =
+        ExchangeClient::new(None, wallet.clone(), Some(BaseUrl::Testnet), None, None)
+            .await
+            .unwrap();
+
+    let max_fee_rate = "10";
+    let builder = "0x1ab189B7801140900C711E458212F9c76F8dAC79";
+
+    info!("something");
```
Comment (lmlmt) [inline src/bin/approve_builder_fee.rs:21]: nit: remove the unnecessary comment

File: src/bin/approve_builder_fee.rs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    let res = exchange_client
+        .approve_builder_fee(builder.to_string(), max_fee_rate.to_string(), Some(&wallet))
+        .await;
+    info!("Response from result: {res:#?}");
```
Comment (lmlmt) [inline src/bin/approve_builder_fee.rs:25]:
```suggestion
    info!("resp: {res:#?}");
```

File: src/bin/approve_builder_fee.rs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    let exchange_client =
+        ExchangeClient::new(None, wallet.clone(), Some(BaseUrl::Testnet), None, None)
+            .await
+            .unwrap();
+
+    let max_fee_rate = "10";
+    let builder = "0x1ab189B7801140900C711E458212F9c76F8dAC79";
+
+    info!("something");
+    let res = exchange_client
```
Comment (lmlmt) [inline src/bin/approve_builder_fee.rs:22]:
```suggestion
    let resp = exchange_client
```

File: src/exchange/actions.rs [hunk 4]
```diff
@@ -293,3 +293,10 @@ pub struct VaultTransfer {
 pub struct SetReferrer {
     pub code: String,
 }
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
```
Comment (lmlmt) [inline src/exchange/actions.rs:297]:
```suggestion
#[derive(Debug, Clone, Serialize, Deserialize)]
```

File: src/exchange/exchange_client.rs [hunk 5]
```diff
@@ -137,16 +138,14 @@ impl ExchangeClient {
...
-
-        serde_json::from_str(
-            &self
-                .http_client
-                .post("/exchange", res)
-                .await
-                .map_err(|e| Error::JsonParse(e.to_string()))?,
-        )
-        .map_err(|e| Error::JsonParse(e.to_string()))
+        info!("Sending request {res:?}");
```
Comment (lmlmt) [inline src/exchange/exchange_client.rs:141]: nit: debug log

**Review by lmlmt — COMMENTED** [review]
Change looks good. Mostly style nits

File: src/exchange/actions.rs (hunk 4 above)
Comment (Frixoe) [reply src/exchange/actions.rs:297]: But every other action has the derive attribute traits in that order. Should I still change it?

File: src/exchange/exchange_client.rs (hunk 5 above)
Comment (Frixoe) [reply src/exchange/exchange_client.rs:141]: done

File: src/bin/approve_builder_fee.rs (hunk 3 above)
Comment (Frixoe) [reply src/bin/approve_builder_fee.rs:22]: done

File: src/bin/approve_builder_fee.rs (hunk 2 above)
Comment (Frixoe) [reply src/bin/approve_builder_fee.rs:25]: done

File: src/bin/approve_builder_fee.rs (hunk 1 above)
Comment (Frixoe) [reply src/bin/approve_builder_fee.rs:21]: done

**Comment by lmlmt** [comment]
This is marked as a draft. Did you test/is it ready to be merged?

**Comment by Frixoe** [comment]
> This is marked as a draft. Did you test/is it ready to be merged?

I'm in the process of adding the `order_with_builder` code too and testing that. Once those are done and tested, I will mark this ready for review and re-request a review.

**Comment by Frixoe** [comment]
Hey @lmlmt, the code is ready for review and to be merged

File: src/exchange/actions.rs [hunk 6]
```diff
@@ -101,6 +101,14 @@ pub struct BulkOrder {
     pub grouping: String,
 }
 
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct BulkOrderWithBuilder {
```
Comment (lmlmt) [inline src/exchange/actions.rs:106]: This doesn't have to be an extra struct. Can reduce copy/paste here and reusability by adding
```
#[serde(default, skip_serializing_if = "Option::is_none")]
pub builder: BuilderInfo,
```
to `BulkOrder`

**Review by lmlmt — COMMENTED** [review]
Mostly looks good, I don't think a separate `BulkOrderWithBuilder` is needed though

**Comment by lmlmt** [comment]
@Frixoe Can you make sure it compiles?

**Comment by Frixoe** [comment]
@lmlmt sorry completely forget about my arch nemesis clippy

**Pull request merged**